feat(developer): line numbers for XML parsing and kmc-ldml 🙀

srl295 opened the pull request:
Fixes: #10622

Updates https://github.com/NaturalIntelligence/fast-xml-parser to 5.2.2

This only adds line numbers to two errors, Error_HardwareLayerHasTooManyRows and Error_KeyMissingToGapOrSwitch - but if the ergonomics look good, we can expand it to the couple hundred call sites easily.

@keymanapp-test-bot skip for now

srl295 commented:
At the error call sites, this is probably a better change:

```diff
-        LdmlCompilerMessages.Error_HardwareLayerHasTooManyRows()
+        LdmlCompilerMessages.Error_HardwareLayerHasTooManyRows(layer)
```

… that is, just add the XML object as a parameter.

srl295 commented:
- now uses a later fast-xml-parser version
- tests fail because the line number symbols are unexpected.

srl295 commented:
- tests now pass

mcdurdin commented on developer/src/common/web/utils/src/xml-utils.ts:
```diff
@@ -123,11 +125,54 @@ const GENERATOR_OPTIONS: KeymanXMLOptionsBag = {
   public constructor(public type: KeymanXMLType) {
   }
 
+  public static offsetToLineColumn(offset: number, text: string) : LineColumn {
+    const lines: string[] = text.split("\n");
+    for (let line = 1; line < lines.length + 1; line++) { // 1-based
+      const linestr = lines[line - 1];
+      if (linestr.length < offset) {
+        offset = offset - (linestr.length + 1); // count newline at end
+        continue;
+      }
+      return { line, column: offset };
+    }
+    return { line: lines.length + 1 };
+  }
```
This is quite inefficient to do for every message; can we do the parse just once when loading the file and store line lengths in a more efficient manner with an R-tree or kd-tree or the like?

srl295 replied:
right. actually, all it probably needs is an array of numbers:  `const lens : number[] = text.split("\n").map(l => l.length+1)`

mcdurdin replied:
> actually, all it probably needs is an array of numbers

The problem with an array is that the whole array has to be scanned to find the matching line. Faster than scanning the file again, but still a lot slower than a R-tree or like which is optimized for nearest-match

srl295 replied:
The current PR only scans one time per file. Could we look at optimizing the lookup separately? I'll encapsulate the lookup separately so that it can be replaced without disruption.

mcdurdin commented on developer/src/common/web/utils/src/compiler-interfaces.ts:
```diff
@@ -5,7 +5,13 @@ import { CompilerCallbacks } from "./compiler-callbacks.js";
   line?: number;
+  /** 
+   * column where a message applies.
+   * where line is unset, column is a file offset
```
I think I'd prefer to have a separate property. They have to be handled differently anyway by other services

srl295 replied:
OK, sounds good. I will leave column in even if we don't expose it at present.

mcdurdin commented on developer/src/common/web/utils/src/xml-utils.ts:
```diff
@@ -8,6 +8,8 @@
 
 import { XMLParser, XMLBuilder } from 'fast-xml-parser';
 
+export const START_INDEX = XMLParser.getStartIndexSymbol();
```
As this turns into a global export on developer-utils, can we name it `XML_START_INDEX_SYMBOL`?

```suggestion
export const XML_START_INDEX_SYMBOL = XMLParser.getStartIndexSymbol();
```

mcdurdin commented on developer/src/kmc/src/util/NodeCompilerCallbacks.ts:
```diff
@@ -154,6 +158,25 @@ export class NodeCompilerCallbacks implements CompilerCallbacks {
+      this.messageFiletext = '';
+    }
+
+    if (event.column && !event.line && event.filename) {
```
I would prefer this plumbing to be done entirely in kmc-ldml: it could pass in `fileColumn`, `xml` to a helper wrapping the `reportMessage` call, e.g. something like:

```ts
reportXmlMessage(callbacks, LdmlCompilerMessages.ERROR_TheXmlIsUnkind(), fileColumn, xmlTree);
```

(Where xmlTree is aforementioned previously compiled R-tree or the like)

This `reportXmlMessage` helper function could exist in developer-utils.

We shouldn't have a direct ref to `LdmlCompilerMessages` outside of `kmc-ldml` (except in the `kmc message` tool).

The thing I don't want is for `NodeCompilerCallbacks`, and all the other callback implementations, to have to do this.

This xml line numbe rreporting will be very beneficial in other XML readers, e.g. KvksFileReader, so we want it to be genericised for future uptake in those places too.

srl295 replied:
thinking about this still.

srl295 replied:
take a look at the latest commit. there's an EventResolver.

The issue is I don't have the XML tree at the error callsite, and I definitely don't have the entire source file content - just 'callbacks' and the specific XML node in question.

mcdurdin replied:
> The issue is I don't have the XML tree at the error callsite, and I definitely don't have the entire source file content - just 'callbacks' and the specific XML node in question.

We should be able to access the XML tree with appropriate design? It's all in a class, and it's cleaner for the data to be cached on the compiler side than on the callback side.

We could even stick it into the LdmlCompilerMessages class as a static. Just keen to try and keep the content handling on the compiler side.

Then, assuming we have a way to get `fileColumn` in the current situation (maybe a helper function, etc), something like this:

```ts
LdmlCompilerMessages.reportXmlMessage(callbacks, LdmlCompilerMessages.ERROR_TheXmlIsUnkind(), fileColumn);
```

```ts
export class LdmlCompilerMessages {
  static xmlTree: any;
  static reportXmlMessage(callbacks: CompilerCallbacks, event: CompilerEvent, column: number) {
    const x = convertColumnToLineNumber(LdmlCompilerMessages.xmlTree, column);
    if(x) {
      event.line = x.line;
      event.column = x.column;
    }
    return event;
  }
}
```

(Again, reportXmlMessage could be in a different module, but the key is to keep too much creeping into the CompilerCallbacks implementations, as these should be kept as clean as possible in terms of data interpretation.)

srl295 replied:
Restructured so that each XML object has the filename it came from, and, the resolving callback (column -> line number) is inserted into the chain by the LDML compiler.

mcdurdin commented on developer/src/kmc/src/util/NodeCompilerCallbacks.ts:
```diff
@@ -154,6 +158,25 @@ export class NodeCompilerCallbacks implements CompilerCallbacks {
+          text = this.messageFiletext;
+        }
+        if (text == '') {
+          text = this.fs.readFileSync(event.filename, "utf-8");
+          if (this.messageFilename == event.filename) {
+            this.messageFiletext = text;
+          }
```
The message processor should not attempt to load source text from disk. It may not have come from disk -- it may be a network resource, memory resource, etc.

srl295 replied:
so the call site down in keys.ts etc doesn't seem to have access to the source XML - maybe it should? 

I think the challenge i see is that the callbacks with reportMessage goes 'directly' back to the consumer, the outer compiler.compile() function (which has the XML source) doesn't get involved.

srl295 replied:
think i've resolved this..

mcdurdin reviewed (changes requested):
Getting in while this is still in draft. I like the concept here but I would like to refactor the design and separation of concerns. Hope this is helpful

srl295 commented on developer/src/common/web/utils/src/xml-utils.ts:
```diff
@@ -184,7 +229,7 @@ export class KeymanXMLReader {
         }
       });
-      return Object.fromEntries(e);
+      return this.copySymbols(Object.fromEntries(e), data);
```
copySymbols/removeSymbols aren't xml specific, would they be better elsewhere (probably still in common/web/utils)?

srl295 commented on developer/src/kmc-ldml/test/helpers/index.ts:
```diff
@@ -327,3 +327,10 @@ const dontEscape = /[a-zA-Z0-9\.${}\[\]-]/;
 }
+
+/** return an object simulating an XML object with a column number */
+export function withColumn(c: number) : any {
+  return {
+    [START_INDEX as any]: c
+  };
+}
```
this also could go in XML utils

srl295 replied:
This is a test util, so leaving it with the test helper.

srl295 commented on developer/src/kmc-ldml/src/compiler/keys.ts:
```diff
@@ -446,7 +446,7 @@ export class KeysCompiler extends SectionCompiler {
     if (layer.row.length > keymap.length) {
       this.callbacks.reportMessage(
-        LdmlCompilerMessages.Error_HardwareLayerHasTooManyRows()
+        LdmlCompilerMessages.Error_HardwareLayerHasTooManyRows(layer)
```
@mcdurdin and all, this is one of the main things i want to review - the ergonomics on the call site for passing in this info.  This way there's no extra helper function invocation

srl295 commented on developer/src/kmc-ldml/src/compiler/ldml-compiler-messages.ts:
```diff
@@ -20,7 +23,8 @@ export class LdmlCompilerMessages {
 
   static ERROR_HardwareLayerHasTooManyRows = SevError | 0x0003;
-  static Error_HardwareLayerHasTooManyRows = () => m(this.ERROR_HardwareLayerHasTooManyRows, `'hardware' layer has too many rows`);
+  static Error_HardwareLayerHasTooManyRows = (x: any) => mx(x,
+    this.ERROR_HardwareLayerHasTooManyRows, `'hardware' layer has too many rows`);
```
@mcdurdin thoughts on this one?

mcdurdin replied:
Looks workable to me. Formatting should match our revised layout for messages, and leave a trailing comma for the missing detail message that we might weant to add in future (trailing comma both suggests another parameter could be added, and makes for simpler code review).

```suggestion
  static Error_HardwareLayerHasTooManyRows = (x: any) => mx(x,
    this.ERROR_HardwareLayerHasTooManyRows, 
    `'hardware' layer has too many rows`,
  );
```

srl295 commented:
> Getting in while this is still in draft. I like the concept here but I would like to refactor the design and separation of concerns. Hope this is helpful

yeah, i wanted to make sure the design and concerns were right first.

srl295 commented:
This now reports err numbers in imported files

```
import-local.xml - info KM05002: Building developer/src/common/web/utils/test/fixtures/ldml-keyboard/import-local.xml
keys-Zyyy-morepunctuation.xml:5 - error KM00011: key id='snail' must have either output=, gap=, or layerId=.
import-local.xml - hint KM00008: Locale 'und' is not minimal or correctly formatted and should be 'en'
import-local.xml - info KM05007: developer/src/common/web/utils/test/fixtures/ldml-keyboard/import-local.xml failed to build.
```

(needs an automated test)

srl295 commented:
@mcdurdin I was restructuring this to track the XML side in the input to the compiler.. but I came across `CompilerFileCallbacks` which seems to be very similar to the approach I'm doing - it's callback-side knowledge of the input file.

srl295 commented:
Anyway, the next move here is to move the LdmlEventResolver out of the callbacks and into the kmc-ldml compiler.

srl295 commented:
this uses statics (global) but is it the right direction?

srl295 reviewed (commented):
@mcdurdin comments
- rename resolver - to generic LineFinder
- remove global resolver - put into compiler context
- lazy parse in the LineFinder
- add LineFinderCache as an object
- map of source files in callback

srl295 commented:
If this passes I'm moving it to review.

Whew.  I ended up just adding the filename to all XML objects’ metadata. Previously, I only put it on _imported_ things (i.e. as an override).  However, it was too challenging to get the "Default" filename out of the CompilerFileCallback and into LDML.

mcdurdin commented on developer/src/kmc-ldml/src/compiler/ldml-compiler-messages.ts:
```diff
@@ -97,8 +105,8 @@ export class LdmlCompilerMessages {
   static ERROR_KeyMissingToGapOrSwitch = SevError | 0x0011;
-  static Error_KeyMissingToGapOrSwitch = (o:{keyId: string}) =>
-  m(this.ERROR_KeyMissingToGapOrSwitch, `key id='${def(o.keyId)}' must have either output=, gap=, or layerId=.`);
+  static Error_KeyMissingToGapOrSwitch = (o:{keyId: string}, x: any) => mx(x,
+    this.ERROR_KeyMissingToGapOrSwitch, `key id='${def(o.keyId)}' must have either output=, gap=, or layerId=.`);
```
```suggestion
  static Error_KeyMissingToGapOrSwitch = (o:{keyId: string}, x: any) => mx(x,
    this.ERROR_KeyMissingToGapOrSwitch, 
    `key id='${def(o.keyId)}' must have either output=, gap=, or layerId=.`,
  );
```

format: Only update to this layout for messages that are modified in this PR. Leave existing messages for a separate format-only PR.

I do wonder if `x:any` should be before `o:...` ergonomically?

srl295 replied:
`kmc/build.sh api` (which i've never run locally..) failed in CI!

```
    const event = cls[f]?.({} /* ignore arguments*/);
```

since 'x' is optional… i think i am going to switch these back

mcdurdin commented on developer/src/kmc/src/commands/buildClasses/BuildProject.ts:
Can you remove this change from the PR?

mcdurdin reviewed (approved):
I think this looks good. Some indenting and trailing whitespace noise (seems your editor has different settings here?)

Does this have any impact on other consumers of XML files -- .kps, .kvks, .kpj? I saw you had to update tests for kpj and kps but no other updates.

Would be good to roll this support into @SabineSIL's work on kmc-convert as well.

srl295 commented on developer/src/common/web/test-helpers/TestCompilerCallbacks.ts:
```diff
@@ -2,7 +2,7 @@ import * as fs from 'node:fs';
 import { CompilerEvent, CompilerCallbacks, CompilerPathCallbacks, CompilerFileSystemCallbacks,
   CompilerError, CompilerNetAsyncCallbacks, DefaultCompilerFileSystemAsyncCallbacks,
-  CompilerFileSystemAsyncCallbacks,
+  CompilerFileSystemAsyncCallbacks, 
```
it sees to have been accidently off. must have been for something else. it's on now..

srl295 commented:
> I think this looks good. Some indenting and trailing whitespace noise (seems your editor has different settings here?)

hopefully settings fixed yet again.

> Does this have any impact on other consumers of XML files -- .kps, .kvks, .kpj? I saw you had to update tests for kpj and kps but no other updates.

I updated the tests to strip out the XML metadata so they don't show up in comparisons. I added a comment explaining why this is being done.  Should not have any impact other than the metadata being present.  So, KPS parsing errs could have line numbers etc - we could hoist the event machinery out of being LDML specific easily, but I think for now having it more contained makes sense.

> Would be good to roll this support into @SabineSIL's work on kmc-convert as well.

:+1:

keyman-server commented:
Changes in this pull request will be available for download in [Keyman version 19.0.36-alpha](https://keyman.com/downloads/releases/alpha/19.0.36)
